client/src/main/java/application/controller/RegisterWindowController.java

```diff
@@ -165,22 +166,74 @@ private boolean fieldsAreValid() {
       return false;
     }
 
-    if (passwordField.getText().isEmpty()) {
+    if (!emailIsValid(emailField.getText())) {
+      errorLabel.setText("Email Address Not Valid");
+      return false;
+    }
+
+    return passwordIsValid(passwordField.getText(), passwordConfirmField.getText());
+  }
+
+  protected Boolean usernameIsValid(String username) {
+    Pattern specialCharPatten = Pattern.compile("[^a-z0-9 ]", Pattern.CASE_INSENSITIVE);
+    Pattern digitCasePatten = Pattern.compile("[0-9 ]");
+    Pattern whiteSpace = Pattern.compile("[\\s]");
+
+    if (username.isEmpty()) {
+      errorLabel.setText("Please Enter Username");
+      return false;
+    }
+
+    if (username.length() > 20) {
+      errorLabel.setText("Username Too Long");
+      return false;
+    }
+
+    if (specialCharPatten.matcher(username).find() || whiteSpace.matcher(username).find()
+    || digitCasePatten.matcher(username).find()) {
+      errorLabel.setText("Username should only contains letters and have no spaces");
+      return false;
+    }
+    return true;
+  }
+
+  protected Boolean emailIsValid(String email) {
+    String regex = "^[\\w-_\\.+]*[\\w-_\\.]\\@([\\w]+\\.)+[\\w]+[\\w]$";
```

Comment:
This is technically too strict, there are some weird emails it'll reject. Do we care?

Reply:
Do you have a regex that would be less strict you think might be better?

Reply:
Email is a fun one to validate, https://haacked.com/archive/2007/08/21/i-knew-how-to-validate-an-email-address-until-i.aspx/ This is a decent write up tho the one he comes up with is a bit wrong due to his understanding of the standard

Reply:
I've changed the regex to:

^[\\w!#$%&'*+/=?`{|}~^-]+(?:\\.[\\w!#$%&'*+/=?`{|}~^-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,6}$

And we can catch anymore with an email confirmation, which were thinking of implementing anyway for checking tutors.